soc/st/stm32/common/stm32_iocell.c

```diff
@@ -0,0 +1,218 @@
+		}                                                                                  \
+	}
+
+static int iocell_init(void)
+{
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	__HAL_RCC_SBS_CLK_ENABLE();
+	/* CSI is required by IO compensation cell */
+	__HAL_RCC_CSI_ENABLE();
+#elif defined(CONFIG_SOC_SERIES_STM32N6X)
+	__HAL_RCC_PWR_CLK_ENABLE();
+	__HAL_RCC_BSEC_CLK_ENABLE();
+	__HAL_RCC_SYSCFG_CLK_ENABLE();
+	/* This implements workaround described in ES0620
+	 * 2.2.21 I/O compensation may deform output signal rise and fall edges
+	 */
+	SYSCFG->VDDCCCR = STM32N6_IOCELL_FIXEDVALUE;
+	SYSCFG->VDDIO2CCCR = STM32N6_IOCELL_FIXEDVALUE;
+	SYSCFG->VDDIO3CCCR = STM32N6_IOCELL_FIXEDVALUE;
+	SYSCFG->VDDIO4CCCR = STM32N6_IOCELL_FIXEDVALUE;
+	SYSCFG->VDDIO5CCCR = STM32N6_IOCELL_FIXEDVALUE;
+#endif
+	DT_FOREACH_CHILD(STM32_IOCELL_NODE, STM32_IOCELL_INIT)
+	return 0;
+}
+
+SYS_INIT(iocell_init, PRE_KERNEL_1, CONFIG_STM32_IOCELL_PRIORITY);
+
+#ifdef CONFIG_LOG
+
+#if defined(CONFIG_SOC_SERIES_STM32N6X)
+#define IOCELL_FUSE_NAME "OTP fuse"
+#else
+#define IOCELL_FUSE_NAME "option bit"
+#endif
+
+#define IOCELL_HSLV_DISCLAIMER                                                                     \
+	"The I/O HSLV configuration must not be enabled "                                          \
+	"if the corresponding power supply is above certain threshold. "                           \
+	"Enabling it while the voltage is higher can damage the device"
+
+static void iocell_hslv_check_log(int domain_id, int safeguard, int hslv, const char *domain_name)
+{
+	if (safeguard < 0) {
+		/* This shouldn't happen with proper implementation */
+		LOG_ERR("Failed to read safeguard value of HSLV: \"%s\" (%d)", domain_name,
+			domain_id);
+	} else if (!safeguard && hslv) {
+		LOG_ERR("HSLV configuration for \"%s\" blocked by " IOCELL_FUSE_NAME, domain_name);
+		LOG_ERR(IOCELL_HSLV_DISCLAIMER);
+	} else if (safeguard && !hslv) {
+		LOG_WRN("HSLV allowed in " IOCELL_FUSE_NAME " for domain \"%s\", but not enabled",
+			domain_name);
+		LOG_WRN(IOCELL_HSLV_DISCLAIMER);
+	}
+}
+
+#define STM32_IOCELL_CHECK(node_id)                                                                \
+	if (STM32_DT_IOCELL_DOMAIN_VALID(DT_PROP(node_id, domain))) {                              \
+		iocell_hslv_check_log(DT_PROP(node_id, domain),                                    \
+				      iocell_get_hslv_safeguard(DT_PROP(node_id, domain)),         \
+				      DT_PROP(node_id, hslv_enabled), DT_NODE_FULL_NAME(node_id)); \
+	} else {                                                                                   \
+		LOG_ERR("Unknown power domain ID: \"" DT_NODE_FULL_NAME(node_id) "\" (%d)",        \
+			DT_PROP(node_id, domain));                                                 \
+	}
+
+static int iocell_check(void){
+	DT_FOREACH_CHILD(STM32_IOCELL_NODE, STM32_IOCELL_CHECK)
+	return 0;
+}
+
+SYS_INIT(iocell_check, POST_KERNEL, CONFIG_STM32_IOCELL_CHECK_PRIORITY);
+#endif
```

Comment:
Please add EOL comments on `#endif` lines - the logic is hard to follow without them:
```suggestion
#endif /* CONFIG_LOG */
```